doc/contributing.rst

```diff
@@ -92,9 +92,14 @@ party packages (like ``numpy`` and ``hyperspy``) and finally kikuchipy imports.
 Making changes
 ==============
 
-Create a new feature branch::
+kikuchipy's git branch model is the Gitflow Workflow (`original blog post
+<https://nvie.com/posts/a-successful-git-branching-model/>`_). This means that if you
+are adding a new feature, branch off of ``develop``, and if you are making a patch
+(hotfix), branch off of `main` instead.
```

Comment:
From reading https://nvie.com/posts/a-successful-git-branching-model, the `main` branch here corresponds to the `release-*` branch in https://nvie.com/posts/a-successful-git-branching-model, which is a bit misleading.

As I understand, `hotfix` is not the correct wording here, I would remove it.

Reply:
I plan to use the `main` branch as nvie's `master` branch. From what I understand from the blog post, the only time we should branch off of this branch is for fixes that should go in a patch release?

The contributing guide explains how to add new features or bug fixes to kikuchipy, while the release guide is in https://github.com/pyxem/kikuchipy/blob/develop/RELEASE.rst.

Agree about "hotfix", I guess the wording would be improved by something like "... and if you are fixing a bug, branch off of `main` instead."

Reply:
> I plan to use the `main` branch as nvie's `master` branch. From what I understand from the blog post, the only time we should branch off of this branch is for fixes that should go in a patch release?

This is not what it says, at least not clearly, instead it says that only `develop` and `hotfix` should branch off `main/master`.
This model strongly recommend that merging to `main/master` trigger a release, from https://nvie.com/posts/a-successful-git-branching-model:

> Therefore, each time when changes are merged back into `master`, this is a new production release by definition. We tend to be very strict at this, so that theoretically, we could use a Git hook script to automatically build and roll-out our software to our production servers everytime there was a commit on `master`.

It looks to me that what is intended here doesn't fit into this model of workflow and therefore I find misleading to use it as a reference. I am not convinced that following this workflow is what is needed neither!

Reply:
OK, I think I understand what you are saying. My plan for bug fixes that should go in a patch release: 
1. Branch off of `main`, do the fix.
2. Here, we can either make release updates directly (changelog, patch number increment) to the bugfix branch, as intended in the blog post, or merge back into `main` and then right away make a patch release branch off of `main`, merge back in, and release a tag to trigger PyPI and eventually the conda-forge bot.

> I am not convinced that following this workflow is what is needed neither!

I agree that it might be a bit overkill, but it is better than how it has been. I've always encountered issues when wanting to merge release updates in a release branch (say `v0.3.x`) back into `master`, because git told me I first had to merge `master` into the release branch before I could do it the other way...

Reply:
> 2\. Here, we can either make release updates directly (changelog, patch number increment) to the bugfix branch, as intended in the blog post, or merge back into `main` and then right away make a patch release branch off of `main`, merge back in, and release a tag to trigger PyPI and eventually the conda-forge bot.

To which branch, should a bug fix PR be made? The `main` or a separate bugfix branch just mentioned above?

Reply:
A bug fix PR should be made to `main`, I believe that should be clear enough from the new wording here: https://github.com/pyxem/kikuchipy/blob/develop/doc/contributing.rst#making-changes

Reply:
Yes, it makes sense, but it goes against what https://nvie.com/posts/a-successful-git-branching-model strongly recommends. For this reason, I find that using this blog post as reference is misleading and can make things even more confusing for new contributors.

Reply:
I agree that it goes against what nvie strongly recommends in the blog post. However, I don't see a problem in that. In my opinion, this way the bug fix is closer to a patch release than if it was on `develop` and had to be cherry-picked into a release branch and then merged into `main`. My motivation for adopting this branching model was to avoid cherry-picking and hopefully release patches more frequently.

I hope that for a new contributor, they will think "OK, I have a bug fix, so let's branch off of main and then make a PR to main". I can then either commit the necessary release updates myself, or merge the PR and make a release branch from/to `main`. After merging these changes to `develop`, everything should be OK.

Reply:
Yes, the approach is fine! What I am saying is that using this blog post as reference in the contributor guide is misleading because it differs significantly from what is actually being done. I think that it would be better to remove it to avoid confusion!

Reply:
> Yes, the approach is fine!

Yes, guess I'm repeating myself here...

> What I am saying is that using this blog post as reference in the contributor guide is misleading because it differs significantly from what is actually being done. I think that it would be better to remove it to avoid confusion!

Hm, I only partly agree. I think this approach is similar enough so that a user who wanted to know more than what is described in the contributing guide, and read the blog post, wouldn't be mislead to make changes off of the wrong branch. As long as that doesn't happen, I would like to keep the reference.

Reply:
> As long as that doesn't happen, I would like to keep the reference.

You have your first case with me, this is what I have asked [above](https://github.com/pyxem/kikuchipy/pull/410#discussion_r692078015)! :) I read the blog post and I understand that bugfixes should not be merged in `main/master` but in another branch.
Note that https://github.com/pyxem/kikuchipy/blob/develop/doc/contributing.rst#making-changes doesn't say to which branch the PR should be done, but at the same time, it is easy to change in the PR itself.
On a similarly topic, the contributor guide doesn't say which branch contains the bug fixes or if both contain bug fixes.

It is a long blog post discussing 5 different types of branch (`main/master`, `hotfixes`, `releases`, `develop`, `feature`), while only 2 are relevant here (`main/master` and `develop`). IMHO it is unnecessarily complicated for what is needed here, particularly for beginners who need to understand how to use branches and PRs.

Reply:
> You have your first case with me, this is what I have asked [above](https://github.com/pyxem/kikuchipy/pull/410#discussion_r692078015)! :)

Haha, you're right! Would just like to say thank you for taking the time to discuss this.

> Note that https://github.com/pyxem/kikuchipy/blob/develop/doc/contributing.rst#making-changes doesn't say to which branch the PR should be done, but at the same time, it is easy to change in the PR itself.

The brief git tutorial in the contributing guide is separated into three parts: "making changes", "keeping your branch up-to-date", and "sharing your changes". It's mentioned in the last part.

> On a similarly topic, the contributor guide doesn't say which branch contains the bug fixes or if both contain bug fixes.

You're right, this is mentioned in the release guide, and should be for maintainers only.

> It is a long blog post discussing 5 different types of branch (`main/master`, `hotfixes`, `releases`, `develop`, `feature`), while only 2 are relevant here (`main/master` and `develop`). IMHO it is unnecessarily complicated for what is needed here, particularly for beginners who need to understand how to use branches and PRs.

You've convinced me! I've removed the reference to the blog post and Gitflow Workflow in the contributing guide, but kept it in the release guide, stating that kikuchipy's approach is *similar* to Gitflow Workflow.